onnxruntime/core/providers/cpu/nn/pool_functors.h

```diff
@@ -377,4 +377,171 @@ struct MaxPool3DTask {
   }
 };
 
+template <typename T>
+struct LpPool1DTask final {
+  const T* X_data;
+  T* Y_data;
+  int64_t x_step;
+  int64_t y_step;
+  int64_t dilation_h;
+  int64_t pooled_height;
+  int64_t stride_h;
+  int64_t height;
+  gsl::span<const int64_t> kernel_shape;
+  gsl::span<const int64_t> pads;
+  int64_t p;
+  TensorOpCost Cost() {
+    double loop_count = static_cast<double>(pooled_height * kernel_shape[0]);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+    for (std::ptrdiff_t c = begin; c < end; ++c) {
+      operator()(c);
+    }
+  }
+  void operator()(std::ptrdiff_t c) const {
+    const T* x_d = X_data + c * x_step;
+    T* y_d = Y_data + c * y_step;
+    for (int64_t ph = 0; ph < pooled_height; ++ph) {
+      int64_t hstart = ph * stride_h - pads[0];
+      int64_t hend = hstart + kernel_shape[0] * dilation_h;
+      y_d[ph] = 0;
+      for (int64_t h = hstart; h < hend; h += dilation_h) {
+        if (math::is_a_ge_zero_and_a_lt_b(h, height)) {
+          y_d[ph] += static_cast<T>(std::pow(std::abs(x_d[h]), p));
+        }
+      }
+      y_d[ph] = static_cast<T>(std::pow(y_d[ph], 1.0f / p));
+    }
+  }
+};
+
+template <typename T>
+struct LpPool2DTask final {
+  const T* X_data;
+  T* Y_data;
+  int64_t x_step;
+  int64_t y_step;
+  int64_t dilation_h;
+  int64_t dilation_w;
+  int64_t pooled_height;
+  int64_t pooled_width;
+  int64_t stride_h;
+  int64_t stride_w;
+  int64_t height;
+  int64_t width;
+  gsl::span<const int64_t> kernel_shape;
+  gsl::span<const int64_t> pads;
+  int64_t p;
+
+  TensorOpCost Cost() {
+    double loop_count = static_cast<double>(pooled_height * pooled_width * kernel_shape[0] * kernel_shape[1]);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+    for (std::ptrdiff_t c = begin; c < end; ++c) {
+      operator()(c);
+    }
+  }
+
+  void operator()(std::ptrdiff_t c) const {
+    const T* x_d = X_data + c * x_step;
+    T* y_d = Y_data + c * y_step;
+    for (int64_t ph = 0; ph < pooled_height; ++ph) {
+      int64_t hstart = ph * stride_h - pads[0];
+      int64_t hend = hstart + kernel_shape[0] * dilation_h;
+      for (int64_t pw = 0; pw < pooled_width; ++pw) {
+        int64_t wstart = pw * stride_w - pads[1];
+        int64_t wend = wstart + kernel_shape[1] * dilation_w;
+        const int64_t pool_index = ph * pooled_width + pw;
+        y_d[pool_index] = 0;
+        for (int64_t h = hstart; h < hend; h += dilation_h) {
+          if (math::is_a_ge_zero_and_a_lt_b(h, height)) {
+            for (int64_t w = wstart; w < wend; w += dilation_w) {
+              if (math::is_a_ge_zero_and_a_lt_b(w, width)) {
+                const int64_t input_index = h * width + w;
+                y_d[pool_index] += static_cast<T>(std::pow(std::abs(x_d[input_index]), p));
```

Comment:
I wonder if is worth create an intermediate index to store `h * width` which does not change in the inner loop.

Reply:
I am concerned with readability here if additional variables are introduced here.  On the hand, compiler will have this computation optimized.